core/src/main/java/edu/wpi/grip/core/AbstractPipelineEntry.java

```diff
@@ -0,0 +1,151 @@
+
+/**
+ * A partial implementation of {@code PipelineEntry} that implements the removal and ID methods.
+ */
+public abstract class AbstractPipelineEntry implements PipelineEntry {
+
+  @VisibleForTesting
+  static final IdPool idPool = IdPool.INSTANCE;
+
+  protected final Object removedLock = new Object();
+  private boolean removed = false;
+  private String id;
+
+  /**
+   * Creates a unique ID string for the given subclass.
+   *
+   * @param entryClass the subclass
+   *
+   * @return an ID string for the subclass.
+   */
+  protected static String makeId(Class<? extends PipelineEntry> entryClass) {
+    String id;
+    do {
+      id = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_HYPHEN, entryClass.getSimpleName())
+          + "." + RandomStringUtils.randomAlphanumeric(8);
+    } while (idPool.checkId(entryClass, id));
+    return id;
+  }
+
+  /**
+   * Creates a new pipeline entry.
+   *
+   * @param id the ID of the new entry. This must be unique for all instances of the concrete class.
+   *
+   * @throws NullPointerException     if the ID is null
+   * @throws IllegalArgumentException if the ID is already taken
+   */
+  protected AbstractPipelineEntry(String id) {
+    idPool.get(getClass()).add(id);
+    this.id = id;
+  }
+
+  @Override
+  public final Socket getSocketByUid(String uid) throws NoSuchElementException {
+    checkNotNull(uid, "UID");
+    for (InputSocket in : getInputSockets()) {
+      if (in.getUid().equals(uid)) {
+        return in;
+      }
+    }
+    for (OutputSocket out : getOutputSockets()) {
+      if (out.getUid().equals(uid)) {
+        return out;
+      }
+    }
+    throw new NoSuchElementException(uid);
+  }
+
+  /**
+   * Cleans up this entry, such as by freeing resources or disabling callbacks.
+   */
+  protected abstract void cleanUp();
+
+  @Override
+  public final void setRemoved() {
+    synchronized (removedLock) {
+      cleanUp();
+      idPool.removeId(this);
+      removed = true;
+    }
+  }
+
+  @Override
+  public final boolean removed() {
+    synchronized (removedLock) {
+      return removed;
+    }
+  }
+
+  @Override
+  public final void setId(String id) throws NullPointerException, IllegalArgumentException {
+    checkNotNull(id, "The ID cannot be null");
+    boolean inDeserialization = Arrays.stream(new Exception().getStackTrace())
+        .map(e -> e.getClassName())
+        .anyMatch(n -> n.matches(".*(Step|Source)Converter"));
+    if (!inDeserialization) {
+      throw new IllegalStateException(
+          "This method may only be called during project deserialization");
+    }
+    idPool.get(getClass()).add(id);
+    this.id = id;
+  }
+
+  @Override
+  public final String getId() {
+    return id;
+  }
+
+  @Override
+  public String toString() {
+    return getId();
+  }
+
+  /**
+   * Pool of used IDs.
+   */
+  @VisibleForTesting
+  static class IdPool extends HashMap<Class<? extends PipelineEntry>, Set<String>> {
```

Comment:
Guava has a type for this:
https://github.com/google/guava/wiki/NewCollectionTypesExplained#implementations